File: results/1.0/ARP/arp-claude-3-5-sonnet-20241022.c
#include <hammer/hammer.h>
#include <stdio.h>
#include <stdlib.h>

// ARP Parser definition
HParser* create_arp_parser(void) {
    // Hardware Type (16 bits)
    HParser* hardware_type = h_uint16();

    // Protocol Type (16 bits)
    HParser* protocol_type = h_uint16();

    // Hardware Address Length (8 bits)
    HParser* hardware_len = h_uint8();

    // Protocol Address Length (8 bits)
    HParser* protocol_len = h_uint8();

    // Operation Code (16 bits)
    HParser* operation = h_uint16();

    // Variable length fields based on hardware_len and protocol_len
    HParser* sender_hardware_addr = h_length_value(hardware_len, h_uint8());
    HParser* sender_protocol_addr = h_length_value(protocol_len, h_uint8());
    HParser* target_hardware_addr = h_length_value(hardware_len, h_uint8());
    HParser* target_protocol_addr = h_length_value(protocol_len, h_uint8());

    // Combine all fields in sequence
    return h_sequence(hardware_type,
                     protocol_type,
                     hardware_len,
                     protocol_len,
                     operation,
                     sender_hardware_addr,
                     sender_protocol_addr,
                     target_hardware_addr,
                     target_protocol_addr,
                     NULL);
}

void print_bytes(uint8_t* data, size_t len) {
    for(size_t i = 0; i < len; i++) {
        printf("%02x", data[i]);
    }
    printf("\n");
}

int main(int argc, char** argv) {
    if(argc != 2) {
        fprintf(stderr, "Usage: %s <arp_packet_file>\n", argv[0]);
        return 1;
    }

    // Read input file
    FILE* f = fopen(argv[1], "rb");
    if(!f) {
        perror("Failed to open file");
        return 1;
    }

    fseek(f, 0, SEEK_END);
    size_t size = ftell(f);
    fseek(f, 0, SEEK_SET);

    uint8_t* input = malloc(size);
    if(fread(input, 1, size, f) != size) {
        perror("Failed to read file");
        fclose(f);
        free(input);
        return 1;
    }
    fclose(f);

    // Create and run parser
    HParser* arp_parser = create_arp_parser();
    HParseResult* result = h_parse(arp_parser, input, size);

    if(!result) {
        fprintf(stderr, "Parse failed\n");
        free(input);
        return 1;
    }

    // Extract and print results
    HParsedToken* token = result->ast;
    if(token->token_type == TT_SEQUENCE) {
        printf("Hardware Type: 0x%04x\n", 
               (unsigned int)token->seq->elements[0]->uint);
        printf("Protocol Type: 0x%04x\n", 
               (unsigned int)token->seq->elements[1]->uint);
        printf("Hardware Address Length: %u\n", 
               (unsigned int)token->seq->elements[2]->uint);
        printf("Protocol Address Length: %u\n", 
               (unsigned int)token->seq->elements[3]->uint);
        printf("Operation: %u\n", 
               (unsigned int)token->seq->elements[4]->uint);
        
        printf("Sender Hardware Address: ");
        print_bytes(token->seq->elements[5]->bytes.token, 
                   token->seq->elements[5]->bytes.len);
        
        printf("Sender Protocol Address: ");
        print_bytes(token->seq->elements[6]->bytes.token, 
                   token->seq->elements[6]->bytes.len);
        
        printf("Target Hardware Address: ");
        print_bytes(token->seq->elements[7]->bytes.token, 
                   token->seq->elements[7]->bytes.len);
        
        printf("Target Protocol Address: ");
        print_bytes(token->seq->elements[8]->bytes.token, 
                   token->seq->elements[8]->bytes.len);
    }

    h_parse_result_free(result);
    free(input);
    return 0;
}

File: results/1.0/ARP/arp-claude-3-5-sonnet-20241022.rs
use nom::combinator::map;
use nom::number::complete::{be_u16, be_u8};
use nom::sequence::tuple;
use nom::{IResult, Parser};
use std::env;
use std::fs::File;
use std::io::Read;

#[derive(Debug)]
pub struct ArpPacket {
    hardware_type: u16,
    protocol_type: u16,
    hardware_addr_len: u8,
    protocol_addr_len: u8,
    operation: u16,
    sender_hardware_addr: Vec<u8>,
    sender_protocol_addr: Vec<u8>,
    target_hardware_addr: Vec<u8>,
    target_protocol_addr: Vec<u8>,
}

fn parse_variable_length(input: &[u8], length: u8) -> IResult<&[u8], Vec<u8>> {
    let (input, bytes) = nom::bytes::complete::take(length as usize)(input)?;
    Ok((input, bytes.to_vec()))
}

fn parse_arp(input: &[u8]) -> IResult<&[u8], ArpPacket> {
    let (input, (hardware_type, protocol_type, hardware_addr_len, protocol_addr_len, operation)) =
        tuple((be_u16, be_u16, be_u8, be_u8, be_u16))(input)?;

    let parser = move |input| {
        let (input, sender_hardware_addr) = parse_variable_length(input, hardware_addr_len)?;
        let (input, sender_protocol_addr) = parse_variable_length(input, protocol_addr_len)?;
        let (input, target_hardware_addr) = parse_variable_length(input, hardware_addr_len)?;
        let (input, target_protocol_addr) = parse_variable_length(input, protocol_addr_len)?;

        Ok((
            input,
            ArpPacket {
                hardware_type,
                protocol_type,
                hardware_addr_len,
                protocol_addr_len,
                operation,
                sender_hardware_addr,
                sender_protocol_addr,
                target_hardware_addr,
                target_protocol_addr,
            },
        ))
    };

    parser(input)
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        eprintln!("Usage: {} <arp_packet_file>", args[0]);
        std::process::exit(1);
    }

    let mut file = File::open(&args[1])?;
    let mut buffer = Vec::new();
    file.read_to_end(&mut buffer)?;

    match parse_arp(&buffer) {
        Ok((remaining, packet)) => {
            println!("Successfully parsed ARP packet:");
            println!("Hardware Type: {:#04x}", packet.hardware_type);
            println!("Protocol Type: {:#04x}", packet.protocol_type);
            println!("Hardware Address Length: {}", packet.hardware_addr_len);
            println!("Protocol Address Length: {}", packet.protocol_addr_len);
            println!("Operation: {}", packet.operation);
            println!(
                "Sender Hardware Address: {:02x?}",
                packet.sender_hardware_addr
            );
            println!(
                "Sender Protocol Address: {:02x?}",
                packet.sender_protocol_addr
            );
            println!(
                "Target Hardware Address: {:02x?}",
                packet.target_hardware_addr
            );
            println!(
                "Target Protocol Address: {:02x?}",
                packet.target_protocol_addr
            );
            
            if !remaining.is_empty() {
                println!("Warning: {} bytes remaining unparsed", remaining.len());
            }
        }
        Err(e) => {
            eprintln!("Failed to parse ARP packet: {:?}", e);
            std::process::exit(1);
        }
    }

    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_parse_arp() {
        // Example ARP packet (28 bytes)
        let test_data = [
            0x00, 0x01, // Hardware type: Ethernet
            0x08, 0x00, // Protocol type: IPv4
            0x06, // Hardware address length
            0x04, // Protocol address length
            0x00, 0x01, // Operation: Request
            0x11, 0x22, 0x33, 0x44, 0x55, 0x66, // Sender hardware address
            0xc0, 0xa8, 0x01, 0x01, // Sender protocol address
            0x00, 0x00, 0x00, 0x00, 0x00, 0x00, // Target hardware address
            0xc0, 0xa8, 0x01, 0x02, // Target protocol address
        ];

        let result = parse_arp(&test_data);
        assert!(result.is_ok());

        let (remaining, packet) = result.unwrap();
        assert!(remaining.is_empty());
        assert_eq!(packet.hardware_type, 1);
        assert_eq!(packet.protocol_type, 0x0800);
        assert_eq!(packet.hardware_addr_len, 6);
        assert_eq!(packet.protocol_addr_len, 4);
        assert_eq!(packet.operation, 1);
    }
}